felix/rules/nat.go

```diff
@@ -151,6 +151,62 @@ func (r *DefaultRuleRenderer) SNATsToIptablesChains(snats map[string]string) []*
 	}}
 }
 
+func (r *DefaultRuleRenderer) EgressSNATsToIptablesChains(snats map[string]string, ipVersion uint8) []*iptables.Chain {
+	if r.Config.BPFEnabled {
+		return r.egressSNATsToIptablesChainsBPF(snats, ipVersion)
+	} else {
+		return r.egressSNATsToIptablesChains(snats, ipVersion)
+	}
+}
+
+func (r *DefaultRuleRenderer) egressSNATsToIptablesChainsBPF(snats map[string]string, ipVersion uint8) []*iptables.Chain {
+	// Extract and sort map keys so we can program rules in a determined order.
+	sortedIntIps := make([]string, 0, len(snats))
+	for intIp := range snats {
+		sortedIntIps = append(sortedIntIps, intIp)
+	}
+	sort.Strings(sortedIntIps)
+
+	rules := []iptables.Rule{}
+	for _, intIp := range sortedIntIps {
+		extIp := snats[intIp]
+		rules = append(rules, iptables.Rule{
+			Match:  iptables.Match().MarkMatchesWithMask(tcdefs.MarkSeenNATOutgoing, tcdefs.MarkSeenNATOutgoingMask),
+			Action: iptables.SNATAction{ToAddr: extIp},
+		})
+	}
+	return []*iptables.Chain{{
+		Name:  ChainNATEgress,
+		Rules: rules,
+	}}
+}
+
+func (r *DefaultRuleRenderer) egressSNATsToIptablesChains(snats map[string]string, ipVersion uint8) []*iptables.Chain {
+	ipConf := r.ipSetConfig(ipVersion)
+
+	allIPsSetName := ipConf.NameForMainIPSet(IPSetIDNATOutgoingAllPools)
+
+	// Extract and sort map keys so we can program rules in a determined order.
+	sortedIntIps := make([]string, 0, len(snats))
+	for intIp := range snats {
+		sortedIntIps = append(sortedIntIps, intIp)
+	}
+	sort.Strings(sortedIntIps)
+
+	rules := []iptables.Rule{}
+	for _, intIp := range sortedIntIps {
+		extIp := snats[intIp]
+		rules = append(rules, iptables.Rule{
+			Match:  iptables.Match().NotDestIPSet(allIPsSetName).SourceNet(intIp),
+			Action: iptables.SNATAction{ToAddr: extIp},
+		})
+	}
+	return []*iptables.Chain{{
+		Name:  ChainNATEgress,
+		Rules: rules,
+	}}
```

Comment:
This seems to be duplicated by egressSNATsToIptablesChainsBPF. What if we just have the if bpf enabled here?

Reply:
Done. I had copied what was done in MakeNatOutgoingRule, but I agree it is better to check BPFEnabled in the same function.